File: hash/top_k_frequent/top_k_frequent.py
# Medium: https://leetcode.com/problems/top-k-frequent-elements/description/


# use hash for count => use list for reversed count, no need to sort
# similar to bucket sort
# complexity: T-O(n), S-O(n)
from collections import defaultdict
def top_k_frequent(nums: list[int], k: int) -> list[int]:
    num_count: dict[int, int] = {}
    for n in nums:
        num_count[n] = num_count.get(n, 0) + 1

    frequency: list[list[int]] = [[] for _ in range(len(nums) + 1)]
    for n, c in num_count.items():
        frequency[c].append(n)

    result: list[int] = []
    for i in range(len(frequency) - 1, 0, -1):
        if frequency[i]:
            result.extend(frequency[i])
        if len(result) >= k:
            return result[:k]
    return result[:k]


# use hash for count => reverse count => sort
# complexity: T-O(nlogn), S-O(n)
def top_k_frequent2(nums: list[int], k: int) -> list[int]:
    num_count: dict[int, int] = {}

    for n in nums:
        num_count[n] = num_count.get(n, 0) + 1

    num_count_reversed: defaultdict[int, list[int]] = defaultdict(list)
    for n, c in num_count.items():
        num_count_reversed[c].append(n)

    count_sorted = sorted(list(num_count_reversed.keys()), reverse=True)

    result: list[int] = []
    i = 0
    while len(result) < k and i < len(count_sorted):
        data: list[int] = num_count_reversed[count_sorted[i]]
        result.extend(data)
        i += 1

    return result[:k]

File: hash/top_k_frequent/top_k_frequent.go
package main

import (
	"slices"
)

// Hash + sort: count the nums, sort by the count, then return the top k elements.
func topKFrequentHashAndSort(nums []int, k int) []int {
	numToCount := make(map[int]int)
	for _, v := range nums {
		numToCount[v]++
	}

	countToNums := make(map[int][]int)
	for num, count := range numToCount {
		countToNums[count] = append(countToNums[count], num)
	}

	counts := []int{}
	for i := range countToNums {
		counts = append(counts, i)
	}

	slices.SortFunc(counts, func(x, y int) int {
		return y - x
	})

	result := []int{}
	i := 0
	for i < len(counts) && len(result) < k {
		result = append(result, countToNums[counts[i]]...)
		i++
	}

	return result
}

// We need to sort by the count, use bucket sort (array index) if the count is not too large.
func topKFrequentBucketSort(nums []int, k int) []int {
	numToCount := make(map[int]int)
	for _, v := range nums {
		numToCount[v]++
	}

	frequency := make([][]int, len(nums)+1)
	for n, c := range numToCount {
		frequency[c] = append(frequency[c], n)
	}

	result := []int{}
	for i := len(frequency) - 1; i > 0; i-- {
		result = append(result, frequency[i]...)
		if len(result) >= k {
			return result
		}
	}

	return result
}
